### PR TITLE
Fix for CVE-2016-6298

### PR DESCRIPTION
**Pull request by simo5**
This fixes the security issue reported in #65

### CODE REVIEW

File: jwcrypto/tests.py [hunk 1]
```diff
@@ -842,6 +842,85 @@ def test_A5(self):
...
+            ok_e = jwe.JWE()
+            ok_e.deserialize(MMA_vector_ok_cek)
+            ko_cek = 0
+            ko_e = jwe.JWE()
+            ko_e.deserialize(MMA_vector_ko_cek)
+
+            import time
+
+            for _ in range(self.iterations):
+                start = time.clock()
```
Comment (tiran) [inline jwcrypto/tests.py]: time.clock() is the worst clock on Unix. You want:

* time.perf_counter() on Python 3.3+
* time.time() on POSIX
* time.clock() on Windows

File: jwcrypto/jwa.py [hunk 2]
```diff
@@ -379,6 +379,23 @@ class _Rsa15(_RSA, JWAAlgorithm):
...
+        self._check_key(key)
+        # Address MMA attack by implementing RFC 3218 - 2.3.2. Random Filling
+        # provides a random cek that will cause the decryption engine to
+        # run to the end, but will fail decryption later.
+
+        # always generate a random cek so we spend roughly the
+        # same time as in the exception side of the branch
+        cek = _randombits(bitsize)
+        try:
+            cek = super(_Rsa15, self).unwrap(key, bitsize, ek, headers)
```
Comment (tiran) [inline jwcrypto/jwa.py]: Is _RSA.unwrap() and private key decryption constant time? Please add comments to all critical paths.

File: jwcrypto/jwa.py [hunk 3]
```diff
@@ -379,6 +379,23 @@ class _Rsa15(_RSA, JWAAlgorithm):
...
+            # always raise so we always run through the exception handling
+            # code in all cases
+            raise Exception('Dummy')
```
Comment (tiran) [inline jwcrypto/jwa.py]: exception chaining and depths of the stack frame may cause different timings. It's fragile.

**Comment by tiran** [comment]
I'm not fully convinced that the mitigation removes all timing variations. It's almost impossible to write constant timing code in Python. Heck, it's even extremely hard in C.

How about we mark PKCS1v15, NULL and other bad algos as insecure and not offer them by default? I have an idea to make the registry more flexible.

File: jwcrypto/tests.py (hunk 1 above)
Comment (simo5) [inline jwcrypto/tests.py]: This is just a tool to get a rough idea of the timing, the variation is still high enough to not really matter I think.

File: jwcrypto/jwa.py (hunk 2 above)
Comment (simo5) [inline jwcrypto/jwa.py]: The openssl implementation of the RSA decrypt function uses contant time code for the critical parts, but it is not constant time in general, and doesn't need to be.

File: jwcrypto/jwa.py (hunk 3 above)
Comment (simo5) [inline jwcrypto/jwa.py]: It does, in my testing the difference between valid CEK and invalid CEKis still measurable in a tight lopp on the local machine, but hardly over the network where the rest of the stack noise is greater than the timing difference. raising an exception there lowered the difference between the code path of about .4% on average in my testing, so I added it. It is not perfect, but can't be,

Note that I measured up to 3% variability in decoding 2 identical messages encrypted with difference CEKs, so as long as we stay in the error margin the attacker has too much noise to figure out what he got.

**Comment by simo5** [comment]
We can't remove all timing variations but that is not the point, we just need to avoid big enough variations to be measurable over the network. Different messages already cause different timings on their own.
It makes no sense to remove RSA1_5 because we have a mitigation and the vulnerability is usage specific, it is only present if an attacker can capture messages and use a server as an oracle. There are perfectly valid uses of this algorithm and it is mandated by the standard,

**Comment by Merenon** [comment]
This patch looks pretty good to me. It's a great idea to always throw an exception: My test cases showed great results with this fix and I believe the difference in timing is negligible over the network.

I agree, that RSA 1.5 is not the best algorithm to support, but as Simo already mentioned, it's necessary to support it in order to be compliant with the specification.
I guess, there are some discussions about deprecating this algorithm though...

Thanks for your fast reaction in fixing this, Simo! It was a pleasure!

**Comment by rokclimb15** [comment]
@simo5 I'm working on backporting the fix to Ubuntu 16.04 (0.2.1). Could you possibly share your tests for the issue so I can write a unit test and confirm the backported fix works?

https://people.canonical.com/~ubuntu-security/cve/2016/CVE-2016-6298.html

**Pull request closed without merging**